enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,349 @@
+configuration parameter to "", but that would affect all images, not just the
+pinned ones. This behavior needs to be changed in CRI-O so that pinned images
+aren't removed, regardless of the value of `version_file_persist`.
+
+The changes to pin the images will be done in a file inside the
+`/etc/crio/crio.conf.d` directory. To avoid potential conflicts with files
+manually created by the administrator the name of this file will be the name of
+the `PinnedImageSet` custom resource concatenated with the UUID assiged by the
+API server. For example, if the custom resource is this:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-pinned-images
+  uuid: 550a1d88-2976-4447-9fc7-b65e457a7f42
+spec:
+  pinnedImages:
+  - quay.io/openshift-release-dev/ocp-release@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  ...
+```
+
+Then the complete path will be this:
+
+```txt
+/etc/crio/crio.conf.d/my-pinned-images-550a1d88-2976-4447-9fc7-b65e457a7f42.conf
+```
+
+The content of the file will the `pinned_images` parameter containing the list
+of images:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+In addition to the list of images to be pinned, the `PinnedImageSet` custom
+resource will also contain a node selector. This is intended to support
+different sets of images for different kinds of nodes. For example, to pin
+different images for control plane and worker nodes the user could create two
+`PinnedImageSet` custom resources:
+
+```yaml
+# For control plane nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/control-plane: ""
+  pinnedImages:
+  ...
+
+---
+
+# For worker nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  pinnedImages:
+  ...
+```
+
+This is specially convenient for pinning images for upgrades: there are many
+images that are needed only by the control plane nodes and there is no need to
+have them consuming disk space in worker nodes.
+
+When no node selector is specified the images will be pinned in all the nodes
+of the cluster.
+
+When a `PinnedImageSet` custom resource is added, modified or deleted the
+machine config operator will create, modify or delete the configuration file,
+reload the CRI-O configuration (with the equivalent of `systemctl reload crio`)
+and then it will use the CRI-O gRPC API to run the equivalent of `crictl pull`
```

Comment:
@jhernand @sinnykumari Any thoughts about what specifically is doing the pulling here? Is that MCD on the node growing a controller, the MCO itself or MCC calling CRI-O remotely over gRPC? I assume this interface isn't already network accessible?

Reply:
I am not very familiar with the machine config code, but I'd say that this should be a new machine-config-controller sub-controller that watches the `PinnedImageSet` custom resources. When it comes to actually pulling the images it should either create a new daemon set  or else let the machine-config-daemon do it, because that task requires access to the gRPC API of CRI-O which is only available via `localhost`. In the [prototype](https://github.com/jhernand/upgrade-tool/blob/main/internal/bundle_loader.go) we used a daemon set.

Reply:
Do you mind updating it to be clear that either the MCD or another new DaemonSet will be responsible for monitoring the PIS configs and retrieving images?

Reply:
I updated the implementation details section to try to clarify that there will be a new _machine-config-controller_ sub controller and a new daemon set. Please check again.

Reply:
I am not sure if we want to create a new deamonset to fetch images. MCO team didn't have any implementation details for it. But per my understanding, how I see this is MCC gains a new sub-controller to process PIS Custom Resource Object supplied to the cluster and renders it into rendered MacineConifg. MCD is where we do usually node level processing. So, if MCD can handle writing PIS related config on node as well as pulling listed  images then we should use MCD. Otherwise, we can look for adding a new daemonset.

Reply:
Sounds good to me. I think this is an implementation detail that we don't need to close in this document.